### ISSUE TITLE
Docker and weave reverse DNS resolution issue.

### ISSUE THREAD

**Issue by amarouni** [weaveworks/weave#2867]
According to #2157 the reverse resolution of a container's IP address, using a weave network, should be fixed by now since the linked Docker PR was merged.

I just wanted to make sure that I'm not missing any new parameter or configuration as I'm still facing the same issue.

Here's my docker :
```
22:49 $ docker version 
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

And weave : 
```
weave version
weave script 1.9.4
weave router 1.9.4
weave proxy  1.9.4
weave plugin 1.9.4
```

And here's my test container : 
```
networks:
  weave:
    external:
      name: weave
services:
  container_0001:
    dns: 172.17.0.1
    dns_search: weave.local.
    hostname: mycnt0001.weave.local
    image: ubuntu:latest
    networks:
    - weave
    privileged: true
    tty: true
version: '2'
```

And from within the container : 
```
ifconfig -a
eth1      Link encap:Ethernet  HWaddr 02:42:AC:13:00:03  
          inet addr:172.19.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe13:3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:96 errors:0 dropped:0 overruns:0 frame:0
          TX packets:16 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:11526 (11.2 KiB)  TX bytes:1212 (1.1 KiB)

ethwe0    Link encap:Ethernet  HWaddr 9A:7A:90:D8:DA:26  
          inet addr:10.2.1.2  Bcast:0.0.0.0  Mask:255.255.255.0
          inet6 addr: fe80::987a:90ff:fed8:da26/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1376  Metric:1
          RX packets:128 errors:0 dropped:0 overruns:0 frame:0
          TX packets:9 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:14556 (14.2 KiB)  TX bytes:690 (690.0 b)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:573 errors:0 dropped:0 overruns:0 frame:0
          TX packets:573 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:76496 (74.7 KiB)  TX bytes:76496 (74.7 KiB)
```

```
cat /etc/resolv.conf 
search weave.local.
nameserver 127.0.0.11
options ndots:0
```

And here's the weave DNS report : 
```
"DNS": {
        "Domain": "weave.local.",
        "Upstream": [
            "127.0.1.1"
        ],
        "Address": "172.17.0.1:53",
        "TTL": 1,
        "Entries": [
            {
                "Hostname": "host.weave.local.",
                "Origin": "26:ed:e2:cc:80:79",
                "ContainerID": "weave:expose",
                "Address": "10.2.1.254",
                "Version": 0,
                "Tombstone": 0
            },
            {
                "Hostname": "mycnt0001.weave.local.",
                "Origin": "26:ed:e2:cc:80:79",
                "ContainerID": "4226d3f29a58ad30b7ed3d09d2511d301998a56e726b440497ec53ca7065fa57",
                "Address": "10.2.1.2",
                "Version": 1,
                "Tombstone": 1490562027
            },
            {
                "Hostname": "mycnt0001.weave.local.",
                "Origin": "26:ed:e2:cc:80:79",
                "ContainerID": "9d82a7b2404a7d479466ecade40079769b09a6566725c97a235507dbc1154b66",
                "Address": "10.2.1.2",
                "Version": 0,
                "Tombstone": 0
            }
        ]
    }
```

And here's the reverse DNS test :
```
dig -x 10.2.1.2
; <<>> DiG 9.8.2rc1-RedHat-9.8.2-0.47.rc1.el6 <<>> -x 10.2.1.2
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 14802
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;2.1.2.10.in-addr.arpa.		IN	PTR

;; ANSWER SECTION:
2.1.2.10.in-addr.arpa.	600	IN	PTR	container_0001.weave.

;; Query time: 0 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Sun Mar 26 23:07:30 2017
;; MSG SIZE  rcvd: 96
```

**Comment by bboreham** [comment]
@amarouni thanks for the extremely comprehensive report. 

In theory, Docker's DNS should ask Weave DNS for the answer, and should not override. I'd like to know where in the chain this breaks down. 

Can you launch Weave Net with `--log-level=debug` and see if you can spot the request/reply in `docker logs weave`, please.

**Comment by amarouni** [comment]
@bboreham I started weave net in debug mode.

The forward DNS requests are correctly resolved and logged by weave : 

`dig mycnt001.weave.local` : 
```
DEBU: 2017/03/30 13:29:47.204352 [nameserver 26:ed:e2:cc:80:79] local request: {MsgHdr:{Id:45462 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:mycnt001.weave.local. Qtype:1 Qclass:1}] Answer:[] Ns:[] Extra:[]}
DEBU: 2017/03/30 13:29:47.204385 [nameserver 26:ed:e2:cc:80:79] lookup mycnt001.weave.local. -> &[10.2.1.1]
DEBU: 2017/03/30 13:29:47.204404 [nameserver 26:ed:e2:cc:80:79] response: ;; opcode: QUERY, status: NOERROR, id: 45462
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;mycnt001.weave.local.	IN	 A

;; ANSWER SECTION:
mycnt001.weave.local.	1	IN	A	10.2.1.1
```

But the reverse DNS requests don't log any thing `dig -x 10.2.1.1`, as if the request is not forwarded to weave. I still get the weird `CONTAINER_NAME.NETWORK_NAME` as response. So why are the reverse requests answered by docker ? Am i missing a new parameter ?

HTH,

**Comment by bboreham** [comment]
Looks like Docker is not even forwarding the request. I don't know of any extra parameter needed. I've been out at CloudNativeCon so not had a chance to experiment. 

Did you ask Docker?

**Comment by amarouni** [comment]
@bboreham looks like the original Docker issue https://github.com/docker/docker/pull/29891 was merged but targeted for 17.04.0 milestone.
Any idea about the Docker PR workflow ?

**Comment by bboreham** [comment]
@amarouni there is a statement at https://github.com/docker/docker/releases/tag/v17.03.0-ce about them being on monthly release cycle now.

There is a release candidate for 17.04: https://github.com/docker/docker/releases/tag/v17.04.0-ce-rc2

**Comment by bboreham** [comment]
Closing on the basis that one needs to use the appropriate version of Docker to get the feature.

**Comment by amarouni** [comment]
@bboreham I'll give the 17.04 a try once it hits GA and report back my findings.
Thanks,

**Comment by amarouni** [comment]
@bboreham The issue was resolved with docker `17.04.0-ce` !
